grammers-crypto/src/aes.rs

```diff
@@ -6,89 +6,84 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+#![allow(deprecated)] // see https://github.com/RustCrypto/block-ciphers/issues/509
+
 //! [AES-Infinite Garble Extension](https://mgp25.com/blog/2015/06/21/AESIGE/) implementation.
 
-#[allow(deprecated)] // see https://github.com/RustCrypto/block-ciphers/issues/509
+use std::mem;
+
+use aes::Aes256;
 use aes::cipher::generic_array::GenericArray;
 use aes::cipher::{BlockDecrypt, BlockEncrypt, KeyInit};
-use std::mem;
 
 /// Encrypt the input plaintext in-place using the AES-IGE mode.
+/// The buffer length must be a multiple of 16.
 pub fn ige_encrypt(buffer: &mut [u8], key: &[u8; 32], iv: &[u8; 32]) {
-    assert!(buffer.len() % 16 == 0);
-
-    #[allow(deprecated)] // see https://github.com/RustCrypto/block-ciphers/issues/509
-    let key = GenericArray::from_slice(key);
-    let cipher = aes::Aes256::new(key);
-
-    let mut plaintext_block = [0; 16];
-    let mut iv1 = [0; 16];
-    let mut iv2 = [0; 16];
-    iv1.copy_from_slice(&iv[..16]);
-    iv2.copy_from_slice(&iv[16..]);
-
-    for ciphertext_block in buffer.chunks_mut(16) {
-        plaintext_block.copy_from_slice(ciphertext_block);
-
-        // block = block XOR iv1
-        ciphertext_block
-            .iter_mut()
-            .zip(plaintext_block)
-            .zip(iv1.as_ref())
-            .for_each(|((x, a), b)| *x = a ^ b);
-
-        // block = encrypt(block);
-        #[allow(deprecated)] // see https://github.com/RustCrypto/block-ciphers/issues/509
-        let ciphertext_block = GenericArray::from_mut_slice(ciphertext_block);
-        cipher.encrypt_block(ciphertext_block);
-
-        // block = block XOR iv2
-        ciphertext_block
-            .iter_mut()
-            .zip(iv2.as_ref())
-            .for_each(|(x, a)| *x ^= a);
-
-        // save ciphertext and adjust iv
-        iv1.copy_from_slice(ciphertext_block);
-        mem::swap(&mut iv2, &mut plaintext_block);
+    assert_eq!(buffer.len() % 16, 0);
+
+    let cipher = Aes256::new(GenericArray::from_slice(key));
+
+    let mut iv1: [u8; 16] = iv[0..16].try_into().unwrap();
+    let mut iv2: [u8; 16] = iv[16..32].try_into().unwrap();
+
+    let mut next_iv2 = [0u8; 16];
+
+    for block in buffer.chunks_mut(16) {
+        // next iv2 = block (plaintext)
+        next_iv2.copy_from_slice(block);
+
+        // block (plaintext) XOR iv1 (previous ciphertext)
+        for i in 0..16 {
+            block[i] ^= iv1[i]
```

Comment:
Pretty sure this was benchmarked long ago to be worse than the zip/for_each. Maybe things have changed though.